docs/devel/infrastructure-interaction.md

```diff
@@ -0,0 +1,85 @@
+**Problem**
+
+We've got at least 3 cases where kube needs to know about and interact with functionality provided by the underlying infrastructure. The 3 I can think of from the top of my head are:
+
+1. minion discovery
+2. storage (be it a GCE PD, NFS, Ceph, or something else [durable data is a little special])
+3. externally reachable functionality (external IPs, external load balancers, ssl termination points, etc)
+
+We don't seem to have a common pattern how kube and the cloud provider/infrastructure should be working together to expose functionality. I'd like to start a discussion of how I see at least a start of how external functionality and kube can get along.
+
+**The three ways kube and infrastructure interact**
+
+1. a new infrastructure ‘function’ is created and a kube resource is automatically created or kube automatically can use this thing (this is how minions work)
+2. a new kube resource is created and the infrastructure satisfies it (this is how pods/service providers work)
+3. a new infrastructure resource is created but kube does not know about it or make use of it until it is told (this is how GCE PDs work)
+
+All three of these have a LOT of utility and value.  At the largest scales 1 and 2 seem likely the right way forward.  But, focussing on 3, right now, is the model that best fits most legacy enterprise workflows and seems to have the most clear way to work across differing infrastructure platforms.  I also believe that a clean separation between infrastructure functionality and kube resources will give us a mental model which makes building the automation toward 1 and 2 easier.  So this document focuses on number 3.  But we would be wise to think about the day when we can define a kube resource and have the infrastructure automatically create the underlying functionality and the day when we can create the functionality in the infrastructure and have it automatically available as a kube resource.
+
+If we attack the problem right now, today, as problem #3 it also leaves the door open for people to write tools on top of kube and their infrastructure to mimic #1 and #2.  Which can hopefully help us one day bring those tools into kube.  But decoupling is a very important first step.
+
+Between minions, GCE PDs, and GCE external load balancers, all of these have different implementations with little coherence for non-gce configurations. So thinking generally, how should kube and infrastructure functionality work.  If I’m looking at the kube cluster from that 3rd infrastructure interaction perspective (and possibly even the 1st perspective) I think we must view that underlying functionality as pets, not cattle.  That’s just a perspective I’m bringing into this discussion.
```

Comment:
I think we should not have any cloud-provider-specific concepts (e.g., GCEPersistentDisk) as first-class primitives in our API. We should have general abstractions and any cloud-specific parameters should be passed via subobject plugins.

I regret I haven't found the time to look at the persistent disk/volume proposals. :-(